doc/step.1.rst
====
step
====

---------------------------------
Breaks up bash scripts into steps
---------------------------------

:Author: blackchip.org
:Date: August 10, 2013
:Copyright: 2013, blackchip.org
:Version: 1.0
:Manual section: 1
:Manual group: Utilities

SYNOPSIS
========

step [options] command...

DESCRIPTION
===========

This can be useful for creating processing steps in which each step
could take a long time to execute. For example, let's say we have the
following script:

    |
    | think --about life
    | think --about universe
    | think --about everything
    | analyze
    | compute_answer

This can all be done by hand, but it is much easier to put this into a
script if it ever needs to be regenerated. The problem is that these
operations can take a long time and it can be easy to make mistakes
while creating this script.

With step, you can do this instead:

    |
    | run life        think --about life
    | run universe    think --about universe
    | run everything  think --about everything
    | run analyze     analyze
    | run answer      compute_answer

Call this script answer.sh. Run the script normally:

    |
    | answer.sh

And all operations are executed in order. But the answer is 56 which
is incorrect. While debugging, there is a problem with the universe
step. Fix it and check its output by only executing that command:

    |
    | step --only universe answer.sh

Once that is working, see if the output from all the thinking looks
okay before computing the answer:

    |
    | step --to everything answer.sh

Now finish it up:

    | 
    | step --from analyze answer.sh

OPTIONS
=======

--banner, -b      Print out a banner before running each step. The
                  banner will be in the format of::
 
                       ===== prog: step

                  where *prog* is the name of the program being exceuted
                  and *step* is the name of the step.
 
--command, -c     Print out the commands of each step as they are
                  executed. If run with the --function option, the
                  x bash option is set while the function is
                  executing.

--from STEP, -f   Start execution of commands starting with *STEP*.

--help, -h        Prints out a summary of usage for this command.

--list, -l        Prints out a list of available steps. This can also
                  used to perform a dry-run of which steps will be
                  executed. 

--only STEP, -o   Only run the specified *STEP* and skip all others.

--skip STEP, -s   Execute all steps execpt for the specified
                  *STEP*. This option can be listed multiple times on
		  the command line to skip more than one step.
 
--to STEP, -t     Stop execution of commands after *STEP*.

--verbose, -v     The same as using the *--banner* and *--command*
                  options.

--version         Prints the version number of this command

NOTES
=====

Scripts to be executed using step should source either /usr/share/step
or /usr/local/share/step depending on the installation location. This
provides the *run* command used for specifying a step.

Step simply controls the conditional execution of *run* commands. All
other commands in the script that do not use *run* are executed
unconditionally. 

Steps that involve more than one command are best placed in a
function:

    |
    | step1() {
    |     echo "command 1"
    |     echo "command 2"
    | }
    | run -f step1

When using *run*, using an *-f* or *--function* option indicates that
the name of the step is also the name of the function to be executed.

EXAMPLES
========

Given the following script, named example.sh:

    |
    | run step1 echo 1
    | run step2 echo 2
    | run step3 echo 3
    | run step4 echo 4

The following prints out "3" and "4":

    |
    | step --from step3 example.sh

The following prints out "1" and "2": 

    |
    | step --to step2 example.sh

The following prints out "2" and "3": 

    |
    | step --from step2 --to step3 example.sh

The following prints out "2" and "4": 

    |
    | step --skip step2 --skip step4 example.sh
 
Dry run the above command with:

    |
    | step --list --skip step2 --skip step4 example.sh


step.lib.sh
#!/bin/bash

# Copyright 2013 blackchip.org
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

__STEP_PROG="$(basename "$0") (step)"

__step_error() {
    echo "$__STEP_PROG: $@" >&2
}

eval set -- "$__STEP_ARGS"
__STEP_SKIPS=""

while true; do 
    case "$1" in
	--banner|-b)
	    __STEP_BANNER=1
	    shift
	    ;;
	--command|-c)
	    __STEP_COMMAND=1
	    shift
	    ;;
	--verbose|-v)
	    __STEP_BANNER=1
	    __STEP_COMMAND=1
	    shift
	    ;;
	--only|-o)
	    __STEP_ONLY=$2
	    __STEP_REQUESTED=$2
	    shift 2
	    ;;
	--from|-f)
	    __STEP_FROM=$2
	    __STEP_DISABLED=1
	    __STEP_REQUESTED=$2
	    shift 2
	    ;;
	--to|-t)
	    __STEP_TO=$2
	    __STEP_REQUESTED=$2
	    shift 2
	    ;;
	--list|-l)
	    __STEP_LIST=1
	    shift
	    ;;
	--skip|-s)
	    __STEP_SKIPS="$__STEP_SKIPS __STEP_SKIP_$2&"
	    shift 2
	    ;;
	--debug|-d)
	    __STEP_DEBUG=1
	    shift
	    ;;
	--)
	    shift
	    break
	    ;;
	*)
	    break
	    ;;
    esac
done

# Banner to be shown before each step is executed when using the --banner
# option. Override by assigning another banner command to the
# STEP_BANNER variable
__step_banner() {
    local prog=$1
    local step=$2

    echo -e "\n===== $prog: $step"
}
STEP_BANNER=__step_banner

# If the --command option is used to print out the commands to be executed,
# this prints out inline commands (function calls set +x). Override by 
# assigning another command to the STEP_COMMAND variable
__step_command() {
    echo "+ $@"
}
STEP_COMMAND=__step_command

if [ "$__STEP_DEBUG" ] ; then
    set -x
fi

__RUN_SHORT_OPTS="f"
__RUN_LONG_OPTS="function"

__run_usage() {
    cat <<EOF
Usage: run [options] <step_name> command...

Options:
    -f, --function      The name of the step is also the name of the function
                        to be executed. 
EOF
}

run() {
    local step=""
    local args=$(POSIXLY_CORRECT=1 getopt -o "$__RUN_SHORT_OPTS" -l "$__RUN_LONG_OPTS" -n "$__STEP_PROG" -- $@ )
    if [ $? -ne 0 ] ; then
	__run_usage
	exit 1
    fi

    eval set -- "$args"
    while true; do
	case "$1" in
	    -f|--function)
                # If this option is given, the name of the step is also the 
		# name of the function to be executed
		local is_function=true
		shift
		;;
	    --)
                shift
		break
		;;
	    *)
		break
		;;
	esac
    done

    step="$1"
    shift
    if [ ! "$step" ] ; then
	__step_error "No step specified in run"
	__run_usage
	exit 1
    fi

    if [ ! "$is_function" ] && [ "$#" -eq 0 ] ; then
	__step_error "No command specified in run"
	__run_usage
	exit 1
    fi

    # Skip if only a specific step is to be run and this is not the step
    if [ "$__STEP_ONLY" ] && [ "$step" != "$__STEP_ONLY" ] ; then
        return 0
    fi

    # Start executing steps if this is the step listed in --from
    if [ "$step" == "$__STEP_FROM" ] ; then
        unset __STEP_DISABLED
    fi

    # Is this a step to be skipped?
    case "$__STEP_SKIPS" in
	*__STEP_SKIP_${step}*)
            __STEP_SKIPPED=1
	    __STEP_PREVIOUS_DISABLED=$__STEP_DISABLED
	    __STEP_DISABLED=1
	    ;;
    esac

    # Exit now if this step should not be run
    if [ "$__STEP_DISABLED" ] ; then
	if [ "$__STEP_SKIPPED" ] ; then
	    __STEP_DISABLED=$__STEP_PREVIOUS_DISABLED
	fi
        return 0
    fi

    __STEP_EXECUTED=1

    # Print banner if requested
    if [ "$__STEP_BANNER" ] ; then
	$STEP_BANNER $(basename "$0") $step
    fi

    # If just listing steps, show it now
    if [ "$__STEP_LIST" ] ; then
	echo "$step"
    else
	# If the --function option was used, the step name is also the
	# function to be executed
	local function=""
	if [ "$is_function" ] ; then
	    function="$step"
	fi

	# If printing commands...
	if [ "$__STEP_COMMAND" ] ; then
	    # If it is a function, let bash do the work
	    if [ "$is_function" ] ; then
		set -x
	    # If inline, print it out
	    else
		$STEP_COMMAND "$@"
	    fi
	fi

	# Execute
	$function "$@" 
	local return_code=$?
	# If --command was set, stop printing commands at this point
	set +x

	# Exit now if the step failed
	if [ $return_code -ne 0 ] ; then
	    exit $return_code
	fi
	if [ "$__STEP_DEBUG" ] ; then
	    # But turn command printing back on if in debug mode
	    set -x
	fi
    fi

    # If this is the final step specified by --to, don't execute any more
    # steps
    if [ "$step" == "$__STEP_TO" ] ; then
        __STEP_DISABLED=1
    fi
    return $return_code
}

# At the end, print out an error message if a step on the command line
# was not found.
__step_check_exit() {
    if [ "$__STEP_REQUESTED" ] && [ ! "$__STEP_EXECUTED" ] ; then
        __step_error "No such step: $__STEP_REQUESTED"
	exit 1
    fi
}

trap __step_check_exit EXIT
